File: harness/platform_core_contract/envelope.py
"""Shared staged Command/Event envelope validation."""

from __future__ import annotations

from typing import Any

from .artifact import (_validate_artifact_references,
                       _validate_artifact_relations,
                       _validate_artifact_values)
from .codec import (ContractError, canonical_json, lower_token, validate_i64,
                    validate_schema_name)
from .constants import (CANONICALIZATION, ENVELOPE_VERSION,
                        DOCUMENT_INVALID, MAX_EXTENSION_FIELDS, MAX_EXTENSIONS_BYTES,
                        MAX_PAYLOAD_BYTES, REFERENCE_MISMATCH,
                        RELATION_MISMATCH)
from .identity import validate_typed_id
from .references import (validate_actor_ref, validate_scope_references,
                         validate_scope_values)


def validate_common_values(node: dict[str, Any]) -> None:
    """Validate shared identifiers and independent field values."""
    if (node["canonicalization"] != CANONICALIZATION or
            validate_i64(node["envelope_version"], "envelope_version", 1)
            != ENVELOPE_VERSION):
        raise ContractError("envelope canonicalization or envelope_version is unsupported")
    validate_i64(node["schema_version"], "schema_version", 1)
    validate_typed_id(node["message_id"], "msg", "message_id")
    validate_typed_id(node["correlation_id"], "cor", "correlation_id")
    if node["causation_id"] is not None:
        validate_typed_id(node["causation_id"], "msg", "causation_id")
    validate_schema_name(node["schema_name"], "schema_name")
    validate_actor_ref(node["actor_ref"])
    validate_scope_values(node["scope_ref"])
    _validate_body_values(node)


def _validate_body_values(node: dict[str, Any]) -> None:
    payload, artifact = node["payload"], node["payload_artifact_ref"]
    if payload is not None:
        canonical_json(payload, MAX_PAYLOAD_BYTES)
    if artifact is not None:
        _validate_artifact_values(artifact)
    _validate_extensions(node["extensions"])


def validate_common_references(node: dict[str, Any]) -> None:
    """Validate scope ancestry and nested Artifact bindings."""
    validate_scope_references(node["scope_ref"])
    artifact = node["payload_artifact_ref"]
    if artifact is None:
        return
    _validate_artifact_references(artifact)
    scope = node["scope_ref"]
    if artifact["source_snapshot_ref"]["entity_id"] != scope["project_snapshot_id"]:
        raise ContractError("payload_artifact_ref source snapshot must match scope",
                            REFERENCE_MISMATCH)
    if artifact["producer_attempt_id"] != scope["attempt_id"]:
        raise ContractError("payload_artifact_ref producer attempt must match scope",
                            REFERENCE_MISMATCH)


def validate_common_relations(node: dict[str, Any]) -> None:
    """Validate shared cross-field relations after all references."""
    if node["causation_id"] == node["message_id"]:
        raise ContractError("causation_id must not equal message_id",
                            RELATION_MISMATCH)
    payload, artifact = node["payload"], node["payload_artifact_ref"]
    if (payload is None) == (artifact is None):
        raise ContractError("exactly one of payload or payload_artifact_ref is required",
                            RELATION_MISMATCH)
    if artifact is not None:
        _validate_artifact_relations(artifact)


def validate_idempotency_key(value: Any) -> None:
    """Validate one bounded visible-ASCII idempotency key."""
    if type(value) is not str or not 16 <= len(value) <= 128:
        raise ContractError("idempotency_key must have 16..128 visible ASCII bytes")
    if any(not "!" <= character <= "~" for character in value):
        raise ContractError("idempotency_key must contain visible ASCII only")


def _validate_extensions(value: dict[str, Any]) -> None:
    if len(value) > MAX_EXTENSION_FIELDS:
        raise ContractError(f"extensions exceeds {MAX_EXTENSION_FIELDS} fields")
    for key in value:
        if type(key) is not str:
            raise ContractError("extension keys must be text", DOCUMENT_INVALID)
        parts = key.split(".")
        if (len(parts) != 2 or not 2 <= len(parts[0]) <= 32 or
                not 1 <= len(parts[1]) <= 64 or
                any(not lower_token(part) for part in parts)):
            raise ContractError(f"extension key {key!r} is not a bounded namespace")
    canonical_json(value, MAX_EXTENSIONS_BYTES)

File: forge-runtime/crates/domain/src/platform_core_contract/envelope.rs
use serde_json::{Map, Value};

use super::{
    ActorRef, ArtifactRef, CANONICALIZATION, ENVELOPE_VERSION, MAX_EXTENSION_FIELDS,
    MAX_EXTENSIONS_BYTES, MAX_PAYLOAD_BYTES, PlatformCoreContractError, RejectionCode, ScopeRef,
    artifact, identity, invalid, references, reject, wire,
};

pub(super) struct EnvelopeView<'a> {
    pub canonicalization: &'a str,
    pub causation_id: &'a Option<String>,
    pub correlation_id: &'a str,
    pub envelope_version: i64,
    pub extensions: &'a Map<String, Value>,
    pub message_id: &'a str,
    pub payload: &'a Option<Map<String, Value>>,
    pub payload_artifact_ref: &'a Option<ArtifactRef>,
    pub schema_name: &'a str,
    pub schema_version: i64,
    pub scope_ref: &'a ScopeRef,
    pub actor_ref: &'a ActorRef,
}

pub(super) fn validate_common_values(
    view: &EnvelopeView<'_>,
) -> Result<(), PlatformCoreContractError> {
    if view.canonicalization != CANONICALIZATION || view.envelope_version != ENVELOPE_VERSION {
        return Err(invalid(
            "envelope canonicalization or envelope_version is unsupported",
        ));
    }
    if view.schema_version < 1 {
        return Err(invalid("schema_version must be positive"));
    }
    validate_common_ids(view)?;
    wire::validate_schema_name(view.schema_name, "schema_name")?;
    references::validate_actor_ref(view.actor_ref)?;
    references::validate_scope_values(view.scope_ref)?;
    validate_body_values(view)
}

fn validate_common_ids(view: &EnvelopeView<'_>) -> Result<(), PlatformCoreContractError> {
    identity::validate_typed_id(view.message_id, "msg", "message_id")?;
    identity::validate_typed_id(view.correlation_id, "cor", "correlation_id")?;
    if let Some(causation_id) = view.causation_id {
        identity::validate_typed_id(causation_id, "msg", "causation_id")?;
    }
    Ok(())
}

fn validate_body_values(view: &EnvelopeView<'_>) -> Result<(), PlatformCoreContractError> {
    if let Some(payload) = view.payload {
        wire::canonical_object(payload, MAX_PAYLOAD_BYTES)
            .map_err(|error| invalid(format!("payload: {error}")))?;
    }
    if let Some(artifact_ref) = view.payload_artifact_ref {
        artifact::validate_artifact_values(artifact_ref)?;
    }
    validate_extensions(view.extensions)
}

pub(super) fn validate_common_references(
    view: &EnvelopeView<'_>,
) -> Result<(), PlatformCoreContractError> {
    references::validate_scope_references(view.scope_ref)?;
    let Some(artifact_ref) = view.payload_artifact_ref else {
        return Ok(());
    };
    artifact::validate_artifact_references(artifact_ref)?;
    if view.scope_ref.project_snapshot_id.as_deref()
        != Some(artifact_ref.source_snapshot_ref.entity_id.as_str())
    {
        return Err(reference_error(
            "payload_artifact_ref source snapshot must match scope",
        ));
    }
    if view.scope_ref.attempt_id.as_deref() != Some(artifact_ref.producer_attempt_id.as_str()) {
        return Err(reference_error(
            "payload_artifact_ref producer attempt must match scope",
        ));
    }
    Ok(())
}

fn reference_error(message: &'static str) -> PlatformCoreContractError {
    reject(RejectionCode::ReferenceMismatch, message)
}

pub(super) fn validate_common_relations(
    view: &EnvelopeView<'_>,
) -> Result<(), PlatformCoreContractError> {
    if view.causation_id.as_deref() == Some(view.message_id) {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "causation_id must not equal message_id",
        ));
    }
    if view.payload.is_some() == view.payload_artifact_ref.is_some() {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "exactly one of payload or payload_artifact_ref is required",
        ));
    }
    if let Some(artifact_ref) = view.payload_artifact_ref {
        artifact::validate_artifact_relations(artifact_ref)?;
    }
    Ok(())
}

fn validate_extensions(value: &Map<String, Value>) -> Result<(), PlatformCoreContractError> {
    if value.len() > MAX_EXTENSION_FIELDS {
        return Err(invalid(format!(
            "extensions exceeds {MAX_EXTENSION_FIELDS} fields"
        )));
    }
    for key in value.keys() {
        validate_extension_key(key)?;
    }
    wire::canonical_object(value, MAX_EXTENSIONS_BYTES)
        .map_err(|error| invalid(format!("extensions: {error}")))?;
    Ok(())
}

fn validate_extension_key(value: &str) -> Result<(), PlatformCoreContractError> {
    let parts: Vec<&str> = value.split('.').collect();
    if parts.len() != 2
        || !(2..=32).contains(&parts[0].len())
        || parts[1].is_empty()
        || parts[1].len() > 64
        || parts.iter().any(|part| !wire::lower_token(part))
    {
        return Err(invalid(format!(
            "extension key {value:?} must have two bounded namespace segments"
        )));
    }
    Ok(())
}

pub(super) fn validate_idempotency_key(value: &str) -> Result<(), PlatformCoreContractError> {
    if !(16..=128).contains(&value.len())
        || !value.bytes().all(|byte| (b'!'..=b'~').contains(&byte))
    {
        return Err(invalid(
            "idempotency_key must contain 16..128 visible ASCII bytes",
        ));
    }
    Ok(())
}
